Source/WebCore/rendering/style/StyleGradientImage.h

```diff
@@ -26,39 +26,92 @@
+        RefPtr<CSSPrimitiveValue> shape;
+        RefPtr<CSSPrimitiveValue> sizingBehavior;
+        RefPtr<CSSPrimitiveValue> endHorizontalSize;
+        RefPtr<CSSPrimitiveValue> endVerticalSize;
+    };
+
+    struct ConicData {
+        RefPtr<CSSPrimitiveValue> firstX;
+        RefPtr<CSSPrimitiveValue> firstY;
+        RefPtr<CSSPrimitiveValue> secondX;
+        RefPtr<CSSPrimitiveValue> secondY;
+        RefPtr<CSSPrimitiveValue> angle;
+    };
+
+    using Data = std::variant<LinearData, RadialData, ConicData>;
+
+    static Ref<StyleGradientImage> create(Data data, CSSGradientRepeat repeat, CSSGradientType gradientType, CSSGradientColorInterpolationMethod colorInterpolationMethod, Vector<Stop> stops)
     {
-        return adoptRef(*new StyleGradientImage(WTFMove(value)));
+        return adoptRef(*new StyleGradientImage(WTFMove(data), repeat, gradientType, colorInterpolationMethod, WTFMove(stops)));
     }
     virtual ~StyleGradientImage();
 
-private:
-    explicit StyleGradientImage(Ref<CSSGradientValue>&&);
-
     bool operator==(const StyleImage&) const final;
+    bool equals(const StyleGradientImage&) const;
+
+    static constexpr bool isFixedSize = false;
+
+private:
+    explicit StyleGradientImage(Data&&, CSSGradientRepeat, CSSGradientType, CSSGradientColorInterpolationMethod, Vector<Stop>&&);
 
-    CSSImageGeneratorValue& imageValue() final;
-    Ref<CSSValue> cssValue() const final;
-    WrappedImagePtr data() const final { return m_imageGeneratorValue.ptr(); }
+    Ref<CSSValue> computedStyleValue(const RenderStyle&) const final;
     bool isPending() const final;
     void load(CachedResourceLoader&, const ResourceLoaderOptions&) final;
     RefPtr<Image> image(const RenderElement*, const FloatSize&) const final;
     bool knownToBeOpaque(const RenderElement&) const final;
     FloatSize fixedSize(const RenderElement&) const final;
+    void didAddClient(RenderElement&) final { }
+    void didRemoveClient(RenderElement&) final { }
+
+    Ref<Gradient> createGradient(const LinearData&, const RenderElement&, const FloatSize&) const;
+    Ref<Gradient> createGradient(const RadialData&, const RenderElement&, const FloatSize&) const;
+    Ref<Gradient> createGradient(const ConicData&, const RenderElement&, const FloatSize&) const;
+
+    template<typename GradientAdapter> GradientColorStops computeStops(GradientAdapter&, const CSSToLengthConversionData&, const RenderStyle&, float maxLengthForRepeat) const;
 
-    void addClient(RenderElement&) final;
-    void removeClient(RenderElement&) final;
-    bool hasClient(RenderElement&) const final;
+    bool hasColorDerivedFromElement() const;
+    bool isCacheable() const;
 
-    Ref<CSSGradientValue> m_imageGeneratorValue;
+    Data m_data;
+    CSSGradientRepeat m_repeat;
+    CSSGradientType m_gradientType;
+    CSSGradientColorInterpolationMethod m_colorInterpolationMethod;
+    Vector<Stop> m_stops;
+    mutable std::optional<bool> m_hasColorDerivedFromElement;
```

Comment:
I thought you were a fervent believer in “no `optional<bool>`”?

Reply:
:( Just one of many things I had to convince myself not to cleanup in this pass (this is just moved from CSSGradientValue). This bool goes away entirely in the next change, so it wasn't worth rocking the boat in this one.